### PR TITLE
gRelayer: CI tests 

### PR DESCRIPTION
**Pull request by derpy-duck**
Gets the following things into CI, as was requested for the Audit (in the same place as all the other CI tests, run as part of sdk-ci-tests)

- integration tests testing the relayer engine and contracts to do a send, forward, multiple sends, multiple forwards, send failure, forward failure, redelivery
- integration tests testing the SDK's ability to read statuses for successes, receiver failures, forward successes, forward failures, and resend from the SDK
- integration tests testing the governance actions of registering a chain, performing an upgrade (thanks Andreas for the debug help), and setting default relay provider

Also significantly cleans up parts of SDK and integration tests

Also addresses bug where no event was emitted when refunds are submitted

### CODE REVIEW

File: sdk/js/src/relayer/__tests__/relay_provider.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    test("readPrices", async () => {
+      const tokenPrice = ethers.BigNumber.from("100000");
+      const gasPrice = ethers.BigNumber.from("300000000000")
+      
+      const tokenPriceReturned = await relayProvider.nativeCurrencyPrice(targetChainId);
+      const gasPriceReturned = await relayProvider.gasPrice(targetChainId);
+
+      expect(tokenPriceReturned.toString()).toBe(tokenPrice.toString());
+      expect(gasPriceReturned.toString()).toBe(gasPrice.toString());
+
```
Comment (nonergodic) [inline sdk/js/src/relayer/__tests__/relay_provider.ts:42]: Equal comparison via the string route is kinda ugly, but I couldn't find a better way of doing it in jest (I would know how to do it in chai...)

File: sdk/js/src/relayer/__tests__/wormhole_relayer.ts [hunk 2]
```diff
@@ -0,0 +1,690 @@
...
+  const env = process.env['ENV'];
+  if(!env) throw Error("No env specified: tilt or ci or testnet or mainnet");
+  const network = env == 'tilt' || env == 'ci' ? "DEVNET" : env == 'testnet' ? "TESTNET" : env == 'mainnet' ? "MAINNET" : undefined;
+  if(!network) throw Error(`Invalid env specified: ${env}`);
+
+  const sourceChainId = network == 'DEVNET' ? 2 : 6;
+  const targetChainId = network == 'DEVNET' ? 4 : 14;
+
+// Devnet Private Key
+const privateKey = process.env['WALLET_KEY'] || "4f3edf983ac636a65a842ce7c78d9aa706d3b113bce9c46f30d7d21715b23b1d"
```
Comment (nonergodic) [inline sdk/js/src/relayer/__tests__/wormhole_relayer.ts:26]: what's that smell...? *sniffs*

It's the smell of ... code duplication!

File: sdk/js/src/relayer/__tests__/wormhole_relayer.ts [hunk 3]
```diff
@@ -0,0 +1,690 @@
...
+  "eb94ff04accbfc8195d44b45e7c7da4c6993b2fbbfc4ef166a7675a905df9891",
+  "053a6527124b309d914a47f5257a995e9b0ad17f14659f90ed42af5e6e262b6a",
+  "3fbf1e46f6da69e62aed5670f279e818889aa7d8f1beb7fd730770fd4f8ea3d7",
+  "53b05697596ba04067e40be8100c9194cbae59c90e7870997de57337497172e9",
+  "4e95cb2ff3f7d5e963631ad85c28b1b79cb370f21c67cbdd4c2ffb0bf664aa06",
+  "01b8c448ce2c1d43cfc5938d3a57086f88e3dc43bb8b08028ecb7a7924f4676f",
+  "1db31a6ba3bcd54d2e8a64f8a2415064265d291593450c6eb7e9a6a986bd9400",
+  "70d8f1c9534a0ab61a020366b831a494057a289441c07be67e4288c44bc6cd5d",
+];
+const GUARDIAN_SET_INDEX = 0;
```
Comment (nonergodic) [inline sdk/js/src/relayer/__tests__/wormhole_relayer.ts:92]: I'm willing to bet that this was copy pasted from somewhere else in the SDK - in which case it should be included instead!

File: sdk/js/src/relayer/__tests__/wormhole_relayer.ts [hunk 4]
```diff
@@ -0,0 +1,690 @@
...
+const governance = new GovernanceEmitter(
+GOVERNANCE_EMITTER_ADDRESS.toBuffer().toString("hex")
+);
+
+const guardianIndices = env=='ci'?[0,1]:[0];
+
+const getStatus = async (txHash: string): Promise<string> => {
+    console.log(env);
+    console.log(sourceChainId);
+    console.log(txHash);
```
Comment (nonergodic) [inline sdk/js/src/relayer/__tests__/wormhole_relayer.ts:113]: seems like a debugging leftover - if not then it should have description of what's being logged

File: sdk/js/src/relayer/__tests__/wormhole_relayer.ts [hunk 5]
```diff
@@ -0,0 +1,690 @@
...
+  walletSource
+);
+const targetCoreRelayer = ethers_contracts.CoreRelayer__factory.connect(
+  targetCoreRelayerAddress,
+  walletTarget
+);
+const targetMockIntegration = ethers_contracts.MockRelayerIntegration__factory.connect(
+  targetMockIntegrationAddress,
+  walletTarget
+);
```
Comment (nonergodic) [inline sdk/js/src/relayer/__tests__/wormhole_relayer.ts:64]: this very much looks like you should have a `source` and a `target` object and a single function to initialize them, seeing how everything here clearly has the same "shape"

(yes it's just testing code, but we should still hold ourselves to some standard...)

File: sdk/js/src/relayer/__tests__/wormhole_relayer.ts [hunk 6]
```diff
@@ -0,0 +1,690 @@
...
+      notEnoughExtraForwardingValue ? 10000 : 800000,
+      await targetCoreRelayer.getDefaultRelayProvider()
+    );
+    console.log(`Quoted gas delivery fee: ${value.add(extraForwardingValue)}`);
+
+    const furtherInstructions: ethers_contracts.MockRelayerIntegration.FurtherInstructionsStruct = {
+      keepSending: true,
+      newMessages: [payload2, "0x00"],
+      chains: [sourceChainId],
+      gasLimits: [500000],
```
Comment (nonergodic) [inline sdk/js/src/relayer/__tests__/wormhole_relayer.ts:159]: I keep seeing 10000 and 500000 pop up in the code - looks like that should be a variable with a name that says what the value is

File: sdk/js/src/relayer/__tests__/wormhole_relayer.ts [hunk 7]
```diff
@@ -0,0 +1,690 @@
...
+    };
+    const tx = await sourceMockIntegration.sendMessagesWithFurtherInstructions(
+      [payload1],
+      furtherInstructions,
+      [targetChainId],
+      [value.add(extraForwardingValue)],
+      { value: value.add(extraForwardingValue), gasLimit: 500000 }
+    );
+    console.log("Sent delivery request!");
+    const rx = await tx.wait();
```
Comment (nonergodic) [inline sdk/js/src/relayer/__tests__/wormhole_relayer.ts:169]: unused variable `rx`

File: sdk/js/src/relayer/__tests__/wormhole_relayer.ts [hunk 8]
```diff
@@ -0,0 +1,690 @@
...
+      targetChainId,
+      500000,
+      await targetCoreRelayer.getDefaultRelayProvider()
+    );
+    const payment = value1
+      .mul(2)
+      .add(value2)
+      .add(value3)
+      .mul(105)
+      .div(100)
```
Comment (nonergodic) [inline sdk/js/src/relayer/__tests__/wormhole_relayer.ts:332]: presumably that's the asset conversion buffer?

File: sdk/js/src/relayer/__tests__/wormhole_relayer.ts [hunk 9]
```diff
@@ -0,0 +1,690 @@
...
+      .add(1);
```
Comment (nonergodic) [inline sdk/js/src/relayer/__tests__/wormhole_relayer.ts:333]: and that's the rounding?

File: sdk/js/src/relayer/__tests__/wormhole_relayer.ts [hunk 10]
```diff
@@ -0,0 +1,690 @@
...
+      `Sent message: ${arbitraryPayload1} (expecting ${arbitraryPayload2} from forward)`
+    );
+    console.log(`Received message on target: ${message1}`);
+    expect(message1).toBe(arbitraryPayload1);
+
+    console.log("Checking if forward message was relayed back");
+    const message2 = await sourceMockIntegration.getMessage();
+    console.log(`Sent message: ${arbitraryPayload2}`);
+    console.log(`Received message on source: ${message2}`);
+    expect(message2).toBe(arbitraryPayload2);
```
Comment (nonergodic) [inline sdk/js/src/relayer/__tests__/wormhole_relayer.ts:242]: numbers in variable names are typically a dead giveaway that the variables in question should instead be an array

(also applies to the rest of the code down below)

**Review by nonergodic — CHANGES_REQUESTED** [review]
Only reviewed the form so far, not the actual function

File: sdk/js/src/relayer/__tests__/wormhole_relayer.ts (hunk 6 above)
Comment (derpy-duck) [reply sdk/js/src/relayer/__tests__/wormhole_relayer.ts]: done

File: sdk/js/src/relayer/__tests__/wormhole_relayer.ts (hunk 7 above)
Comment (derpy-duck) [reply sdk/js/src/relayer/__tests__/wormhole_relayer.ts:127]: resolved

File: sdk/js/src/relayer/__tests__/wormhole_relayer.ts (hunk 8 above)
Comment (derpy-duck) [reply sdk/js/src/relayer/__tests__/wormhole_relayer.ts]: Yes, added comment

File: sdk/js/src/relayer/__tests__/wormhole_relayer.ts (hunk 9 above)
Comment (derpy-duck) [reply sdk/js/src/relayer/__tests__/wormhole_relayer.ts]: Yup!

File: sdk/js/src/relayer/consts.ts [hunk 11]
```diff
@@ -1,74 +1,55 @@
...
 import { CoreRelayer__factory, CoreRelayer } from "../ethers-contracts/";
+import { number } from "yargs";
 
-const TESTNET = [
-  {
-    chainId: 4,
+type AddressInfo = {coreRelayerAddress?: string, mockRelayProviderAddress?: string, mockIntegrationAddress?: string}
+
+const TESTNET: {[chainId: number]: AddressInfo} = {
+  4: {
```
[the remark this answers is not in the document]
Comment (derpy-duck) [reply sdk/js/src/relayer/consts.ts]: Agreed, made this change!

File: sdk/js/src/mock/governance.ts [hunk 12]
```diff
@@ -202,4 +202,69 @@ export class GovernanceEmitter extends MockEmitter {
       uptickSequence
     );
   }
+
+  publishWormholeRelayerRegisterChain(
+    timestamp: number,
+    chain: number,
```
[the remark this answers is not in the document]
Comment (derpy-duck) [reply sdk/js/src/mock/governance.ts:208]: Agreed, made this change!

File: sdk/js/src/relayer/__tests__/relay_provider.ts [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+  const oracleDeployer = new ethers.Wallet(privateKey, provider);
+  const relayProviderAddress = addressInfo.mockRelayProviderAddress;
+  if(!relayProviderAddress) throw Error("No relay provider address");
+  const relayProvider = RelayProvider__factory.connect(relayProviderAddress, oracleDeployer);
+
+
+  describe("Read Prices Correctly", () => {
+    test("readPrices", async () => {
+      const tokenPrice = ethers.BigNumber.from("100000");
+      const gasPrice = ethers.BigNumber.from("300000000000")
```
[the remark this answers is not in the document]
Comment (derpy-duck) [reply sdk/js/src/relayer/__tests__/relay_provider.ts]: done!

File: sdk/js/src/relayer/__tests__/relay_provider.ts (hunk 1 above)
Comment (derpy-duck) [reply sdk/js/src/relayer/__tests__/relay_provider.ts:42]: :'(

File: sdk/js/src/relayer/__tests__/wormhole_relayer.ts (hunk 2 above)
Comment (derpy-duck) [reply sdk/js/src/relayer/__tests__/wormhole_relayer.ts]: Removed :)

File: sdk/js/src/relayer/__tests__/wormhole_relayer.ts (hunk 3 above)
Comment (derpy-duck) [reply sdk/js/src/relayer/__tests__/wormhole_relayer.ts]: It actually wasn't; agree that this should be done better but this is the first instance in the SDK (it came from the testing/solana-test-validator folder).

I put a comment in the code about how this should likely live in the SDK as the only location, and the testing folder should get it from there.

Comment (derpy-duck) [reply sdk/js/src/relayer/__tests__/wormhole_relayer.ts]: Also moved it to the utils file

File: sdk/js/src/relayer/__tests__/wormhole_relayer.ts (hunk 4 above)
Comment (derpy-duck) [reply sdk/js/src/relayer/__tests__/wormhole_relayer.ts]: yup oops

File: sdk/js/src/relayer/__tests__/relay_provider.ts [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+import {getDefaultProvider} from "../main/helpers"
+
+
+const env = process.env['ENV'];
+if(!env) throw Error("No env specified: tilt or ci or testnet or mainnet");
+const network = env == 'tilt' || env == 'ci' ? "DEVNET" : env == 'testnet' ? "TESTNET" : env == 'mainnet' ? "MAINNET" : undefined;
+if(!network) throw Error(`Invalid env specified: ${env}`);
+
+const sourceChainId = network == 'DEVNET' ? 2 : 6;
+const targetChainId = network == 'DEVNET' ? 4 : 14;
```
[the remark this answers is not in the document]
Comment (derpy-duck) [reply sdk/js/src/relayer/__tests__/relay_provider.ts]: agreed, done!

File: sdk/js/src/relayer/__tests__/wormhole_relayer.ts (hunk 5 above)
Comment (derpy-duck) [reply sdk/js/src/relayer/__tests__/wormhole_relayer.ts]: agreed, done!

**Review by chase-45 — APPROVED** [review]

**Review by nonergodic — APPROVED** [review]

**Pull request merged**